File: test/integ/ConstructorDedupTest.java
/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.redextest;

class ConstructorDedupTest {
  public int f0;
  public boolean f1;
  public Object f2;
  public Object f3;
  public Object f4;

  static class MergeableType1 {}

  static class MergeableType2 {}

  static class MergeableType3 {}

  static class MergeableType4 {}

  public ConstructorDedupTest(MergeableType1 v0, MergeableType2 v1, int v2) {
    f2 = v0;
    f0 = v2;
    f3 = v1;
  }

  public ConstructorDedupTest(MergeableType3 v0, MergeableType4 v1, int v2) {
    f2 = v0;
    f0 = v2;
    f3 = v1;
  }

  public ConstructorDedupTest(MergeableType1 v0, MergeableType3 v1, int v2) {
    f2 = v0;
    f0 = v2;
    f4 = v1;
  }

  public ConstructorDedupTest(MergeableType3 v0, MergeableType2 v1, int v2) {
    f2 = v0;
    f0 = v2;
    f4 = v1;
  }

  public ConstructorDedupTest(MergeableType3 v0, boolean v1, int v2) {
    f2 = v0;
    f0 = v2;
    f1 = v1;
  }

  public ConstructorDedupTest(MergeableType4 v0, boolean v1, int v2) {
    f2 = v0;
    f0 = v2;
    f1 = v1;
  }

  public ConstructorDedupTest(int v0, boolean v1, Integer v2) {
    f0 = v0;
    f2 = v2;
    f1 = v1;
  }

  public ConstructorDedupTest(String v2, int v0, boolean v1) {
    f1 = v1;
    f2 = v2;
    f0 = v0;
  }

  public ConstructorDedupTest(boolean v1, int v0, String v2) {
    f0 = v0;
    f1 = v1;
    f2 = v2;
  }

  // The two constructors are the same but constructors with unused argument is not currently supported.
  public ConstructorDedupTest(boolean v1, String v2, int v0, int v3) {
    f0 = v0;
    f1 = v1;
    f2 = v2;
  }

  public ConstructorDedupTest(String v2, boolean v1, int v0, int v3) {
    f0 = v0;
    f2 = v2;
    f1 = v1;
  }

  public static void dedup_0() {
    ConstructorDedupTest obj1 = new ConstructorDedupTest(1, true, Integer.valueOf(1));
    ConstructorDedupTest obj2 = new ConstructorDedupTest("O", 2, false);
    ConstructorDedupTest obj3 = new ConstructorDedupTest(true, 3, "N");
  }

  public static void dedup_1(
      MergeableType1 mt1, MergeableType2 mt2, MergeableType3 mt3, MergeableType4 mt4) {
    ConstructorDedupTest obj1 = new ConstructorDedupTest(mt1, mt2, 5);
    ConstructorDedupTest obj2 = new ConstructorDedupTest(mt3, mt4, 10);
  }

  public static void dedup_2(MergeableType1 mt1, MergeableType2 mt2, MergeableType3 mt3) {
    ConstructorDedupTest obj1 = new ConstructorDedupTest(mt1, mt3, 5);
    ConstructorDedupTest obj2 = new ConstructorDedupTest(mt3, mt2, 10);
  }

  public static void not_supported() {
    ConstructorDedupTest obj4 = new ConstructorDedupTest(true, "N", 4, 4);
    ConstructorDedupTest obj5 = new ConstructorDedupTest("N", true, 4, 4);
  }
}

File: test/integ/ConstructorDedupTest.cpp
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

#include "DexClass.h"
#include "IRTypeChecker.h"
#include "NormalizeConstructor.h"
#include "RedexTest.h"
#include "Show.h"
#include "Walkers.h"

class ConstructorDedupTest : public RedexIntegrationTest {};

TEST_F(ConstructorDedupTest, dedup) {
  auto type =
      DexType::get_type("Lcom/facebook/redextest/ConstructorDedupTest;");
  auto cls = type_class(type);
  EXPECT_NE(cls, nullptr);
  std::vector<DexClass*> scope({cls});
  auto ctors = cls->get_ctors();
  auto dedupped = method_dedup::dedup_constructors(scope, scope);
  EXPECT_EQ(dedupped, 6);
  walk::parallel::methods(scope, [&](DexMethod* method) {
    IRTypeChecker checker(method);
    checker.run();
    if (checker.fail()) {
      std::string msg = checker.what();
      fprintf(stderr, "ABORT! Inconsistency found in Dex code for %s.\n %s\n",
              SHOW(method), msg.c_str());
      fprintf(stderr, "Code:\n%s\n", SHOW(method->get_code()->cfg()));
      exit(EXIT_FAILURE);
    }
    auto method_name = method->str();
    if (method_name.find("dedup_") == std::string::npos) {
      return;
    }
    DexMethod* ctor{nullptr};
    if (method_name == "dedup_0") {
      // All the constructor invocations are calling ctors[0].
      ctor = ctors[0];
    } else if (method_name == "dedup_1") {
      // All the constructor invocations are calling ctors[1] - the proto also
      // contains additional integer parameters to address collisions
      ctor = ctors[1];
    } else {
      // All the constructor invocations are calling ctors[2].
      ctor = ctors[2];
    }

    for (auto& mie : InstructionIterable(method->get_code())) {
      auto insn = mie.insn;
      if (insn->has_method()) {
        auto callee = insn->get_method();
        if (callee->get_class() == type && method::is_init(callee)) {
          // Only one constructor is used.
          EXPECT_EQ(callee, ctor);
        }
      }
    }
  });
}
